### ISSUE TITLE
Respect oidc_login_next url in OIDCAuthenticationCallbackView even on failure

### ISSUE THREAD

**Issue by jaap3** [mozilla/mozilla-django-oidc#458]
For the longest time we've had this "bug" in our application where users are redirected to the homepage when they return after the weekend. I was tasked to fix this and finally tracked it down to mozilla-django-oidc's `SessionRefresh` middleware.

What happens is the following.

* User is logged in using OIDC
* User goes away for a while
* User comes back, refreshes the page (or manually navigates to "some page")
* The `SessionRefresh` middleware checks if the OIDC session is still valid and sets `session['oidc_login_next']` to return the user to the original location.
* The users session on the OIDC server has expired, resulting in a error state at the `OIDCAuthenticationCallbackView`
* User is redirected to /

I've patched this in our application by subclassing `OIDCAuthenticationCallbackView`:

```
@method_decorator(never_cache, name='dispatch')
class MyOIDCAuthenticationCallbackView(OIDCAuthenticationCallbackView):
    def dispatch(self, request, *args, **kwargs):
        # Store next_url, on failure the session is flushed and this value will be lost.
        self.next_url = request.session.get('oidc_login_next', None)
        return super().dispatch(request, *args, **kwargs)

    @property
    def failure_url(self):
        # Changed from base implementation to respect the oidc_login_next
        # value and redirect back to the original (potentially protected) view
        # even if session refresh fails.
        return getattr(self, 'next_url', None) or super().failure_url
```

Is this something that would be accepted if I made a PR? Or is the current functionality an intentional design choice?

**Comment by akatsoulas** [comment]
@jaap3 that's a good question. I believe there are valid arguments for both cases. @willkg thoughts?

**Comment by willkg** [comment]
@jaap3 What error state in `OIDCAuthenticationCallbackView` are we talking about? Can you point to the code in question?

**Comment by jaap3** [comment]
@willkg I'm talking about the following: 

A user is logged in (according to their Django session), but has been idle long enough for the `id_token` to expire. When the user becomes active again the following happens on their first request:

The `SessionRefresh` middleware tries to silently re-aquire a `id_token` (and saves the current request url in the session in order to return the user to their intended location afterward): https://github.com/mozilla/mozilla-django-oidc/blob/829d2ed67b182c3ef7ee92e7080aaec45f41f911/mozilla_django_oidc/middleware.py#L139-L161

The authentication server cannot complete the OIDC request as the user's session has expired on their side. Since the `prompt` param is set to `none` it must not display any authentication or consent screen. So the server redirects the user to the `redirect_uri` with a error param (according to the OIDC spec). In my case it's `/oidc/callback/?error=login_require&state=...`, but it may also be some of the other ones mentioned here: https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.3.1.2.6

Now the first `if` block of `OIDCAuthenticationCallbackView.get` kicks in and the user is logged out. (`auth.logout(request)` resets the session, so at this point `request.session['oidc_login_next']` is not available anymore): https://github.com/mozilla/mozilla-django-oidc/blob/829d2ed67b182c3ef7ee92e7080aaec45f41f911/mozilla_django_oidc/views.py#L66-L82

Finally the user is redirected to `self.failure_url` is, which by default is `/`:
https://github.com/mozilla/mozilla-django-oidc/blob/829d2ed67b182c3ef7ee92e7080aaec45f41f911/mozilla_django_oidc/views.py#L35-L37.

All of this is invisible to the user, so the end result is a vague report of "This morning I clicked a link in /admin/ and somehow ended up at the homepage" that is almost impossible to reproduce.

What I'd like to see is that, even in the case of a silent reauthentication error, the user is redirected back to their intended location. This might be an unprotected view (so no login is required, the user is just silently logged out), or a protected view. In the latter case the user is redirect back to the authentication server, but this time the `prompt` param is absent so they are able to login again and will still end up at their intended location.

I hope this clearly describes the case.

**Comment by akatsoulas** [comment]
Figuring out what's the proper design, it's an interesting question here. I believe the main problem is the loss of information along the way when there is an `error`. I am leaning towards in updating the code-base to preserve the `oidc_login_next` but without changing the behavior of the `failure_url`. I do see some value to have that property configurable through a setting by default and by updating it we will be limiting that ability. By preserving the information from the session, it's really easy for anyone to override that setting to always respect the `next_url` when present. 

Having said that, @willkg may have a better idea on how to approach this

**Comment by willkg** [comment]
That's a lot clearer--thank you!

The solution you arrived at in the description looks like it's taking the value of `oidc_login_next` from the request session (which is unique to each user) and putting it in the `MyOIDCAuthenticationCallbackView` session instance attribute which is global to all requests running on that worker. Does that sound right? Is there some session-local thing I'm not seeing going on?

**Comment by jaap3** [comment]
@willkg luckily Django's class based views instantiated per-request, that's what the `as_view` call is for in https://github.com/mozilla/mozilla-django-oidc/blob/829d2ed67b182c3ef7ee92e7080aaec45f41f911/mozilla_django_oidc/urls.py#L21. There's no process wide instance that gets to process all requests as that would be a giant footgun as you described.

See also https://github.com/django/django/blob/292f372768836e2aebc713064c5139e8067eebcb/django/views/generic/base.py#L79-L122

**Comment by willkg** [comment]
@jaap3 Ahh... that makes sense. Thank you! Does using Django's `user_logged_in` and `user_logged_out` signals help you here?

**Comment by willkg** [comment]
Fair enough. I'll find some time to throw together a PR and maybe that'll make it clearer.
